### PR TITLE
JESI-4588

### PR DESCRIPTION
**Pull request by AndreTheHunter**
[JESI-4588](https://jesims.atlassian.net/browse/JESI-4588)
- [x] Release https://github.com/jesims/parent/pull/65
- [x] Fix cljs compiler issue
- [x] Release https://github.com/jesims/parent/pull/68

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -8,12 +8,12 @@
             :distribution :repo
             :comments     "same as Clojure"}
   :plugins [[lein-parent "0.3.8"]]
-  :parent-project {:coords  [io.jesi/parent "4.5.0"]
+  :parent-project {:coords  [io.jesi/parent "4.11.0"]
```
Comment (AndreTheHunter) [inline project.clj:11]: https://github.com/jesims/parent/pull/65

File: src/io/jesi/backpack/collection.cljc [hunk 2]
```diff
@@ -365,3 +365,10 @@
   (if (empty? x)
     nil
     x))
+
+(defn distinct-vals?
+  "Returns true if all the map values are unique "
+  [m]
+  ;TODO this the most efficient way?
```
Comment (AndreTheHunter) [inline src/io/jesi/backpack/collection.cljc:372]: Any ideas?

File: src/io/jesi/backpack/macros.cljc [hunk 3]
```diff
@@ -220,3 +222,64 @@ A single default expression can follow the clauses, and its value will be return
...
+        ~@(for [[k v] (partition 2 kvs)]
+            `(assoc-nx ~k ~v)))))
+
+(defmacro atom-assoc!
+  "Lazily `atom/assoc!` the `body` if key `k` does not already exist in the atom.
+  Returns the `assoc`ed (`body`) value"
+  [a k body]
+  ;;TODO make this 100% thread-safe
+  ;;currently body could be evaluated concurrently between the @ and the swap!
+  ;;create own Atom implementation that does lock
```
Comment (AndreTheHunter) [inline src/io/jesi/backpack/macros.cljc:281]: I don't think it's working doing that now.

Thoughts?

File: test/io/jesi/backpack/macros_test.cljc [hunk 4]
```diff
@@ -272,3 +274,100 @@
...
+    (testing "on exception"
+      (let [o (->Closey)]
+        (is (thrown? ExceptionInfo (with-open [o o]
+                                     (when (is (false? (closed? o)))
+                                       (throw (ex-info "Don't panic" {}))))))
+        (is (true? (closed? o)))))))
+
+;FIXME complete
+(deftest with-open->-test)
+
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/macros_test.cljc:297]: 🎗

File: test/io/jesi/backpack/macros_test.cljc [hunk 5]
```diff
@@ -272,3 +274,100 @@
...
+        (is (identical? m2
+                        (assoc-nx m2
+                          :a (throw (ex-info "Unexpected a" {}))
+                          :b (throw (ex-info "Unexpected b" {})))
+                        (assoc-nx m2
+                          [:a (throw (ex-info "Unexpected a" {}))
+                           :b (throw (ex-info "Unexpected b" {}))])))))))
+
+;FIXME complete
+(deftest atom-assoc!-test)
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/macros_test.cljc:373]: 🎗

File: test/io/jesi/backpack/macros_test.cljc [hunk 6]
```diff
@@ -272,3 +274,100 @@
...
+        m2 (assoc m1
+             :b 2)
+        m3 (assoc m2
+             :c 3)]
+
+    #?(:clj (testing "is a macro"
+              (is (macro? `assoc-nx))))
+
+    (testing "expands"
+      ;Fails even though the content is the same (yeah that again)
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/macros_test.cljc:318]: I'm getting sick of this issue. Will need ✋🧠👀

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,15 @@
...
+
+Added:
+
+* `io.jesi.backpack.macros/`:
+  * `with-open`
+  * `with-open->`
+  * `assoc-nx`
+  * `lazy-assoc!`
+* `io.jesi.backpack.close/close` multimethod
+* `io.jesi.backpack.collection/distinct-vals?`
```
Comment (AndreTheHunter) [inline CHANGELOG.md:11]: I did some stuff

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,16 @@
...
+
+* `io.jesi.backpack.macros/`:
+  * `with-open`
+  * `with-open->`
+  * `assoc-nx`
+  * `assoc-nx!`
+* `io.jesi.backpack.close/close` multimethod
+* `io.jesi.backpack.collection/distinct-vals?`
+* `io.jesi.backpack.miscellaneous/re-quote`
+
```
Comment (AndreTheHunter) [inline CHANGELOG.md:13]: I did stuff

File: src/io/jesi/backpack/macros.cljc [hunk 9]
```diff
@@ -220,3 +222,65 @@ A single default expression can follow the clauses, and its value will be return
...
+        ~@(for [[k v] (partition 2 kvs)]
+            `(assoc-nx ~k ~v)))))
+
+(defmacro assoc-nx!
+  "Lazily `atom/assoc!` if the key does not already exist.
+  Returns the value in the atom"
+  [a k v]
+  ;;TODO make this 100% thread-safe
+  ;;currently body could be evaluated concurrently between the @ and the swap! (atom/assoc!)
+  ;;create own Atom implementation that does lock
```
Comment (AndreTheHunter) [inline src/io/jesi/backpack/macros.cljc:281]: I don't think it's worth doing now. Thoughts?

File: test/io/jesi/backpack/clojure_test.clj [hunk 10]
```diff
@@ -12,13 +12,17 @@
 (def-test-type ::and-worked false)
 (def-test-type :LOTS_OF_YELLING true)
 
+(defn- class-name [o]
+  (-> o (class) (.getName)))
+
 (deftest defkw-type-test
 
   (testing "Registers keyword as type"
-    ;TODO fails in test refresh (but not the first time it runs)
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/clojure_test.clj:18]: Got sick of this

File: test/io/jesi/backpack/miscellaneous_test.cljc [hunk 11]
```diff
@@ -93,3 +95,17 @@
...
     (is (false? (bp/xor true false true)))))
+
+(deftest re-quote-test
+
+  (spy/enabled
+    (testing "quotes the regex string"
+      (let [s ".*+?^${}()|[]\\"]
+        #?(:clj  (is= (str "\\Q" s "\\E")
+                      (bp/re-quote s))
+           ;FIXME check the string, there should be a \\ before each character
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/miscellaneous_test.cljc:106]: 🎗

File: test/io/jesi/backpack/miscellaneous_test.cljc [hunk 12]
```diff
@@ -2,7 +2,9 @@
   (:require
     [io.jesi.backpack :as bp]
     [io.jesi.backpack.random :as rnd]
-    [io.jesi.customs.strict :refer [are deftest is is= testing]])
+    [io.jesi.customs.spy :as spy]
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/miscellaneous_test.cljc:5]:
```suggestion
```

File: project.clj [hunk 13]
```diff
@@ -8,12 +8,20 @@
...
                    :inherit [:plugins :managed-dependencies :deploy-repositories :dependencies :profiles :test-refresh :aliases :codox]}
-  :managed-dependencies [[io.jesi/backpack "6.3.0"]]
+  :managed-dependencies [;[io.jesi/backpack "6.3.0"]
+                         [io.jesi/backpack "7.1.0"]]
+                         ;FIXME once CLJS is working, update in parent
```
Comment (AndreTheHunter) [inline project.clj:15]: 🎗

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,3 +1,16 @@
...
+Added:
+
+* `io.jesi.backpack.macros/`:
+  * `with-open`
+  * `with-open->`
+  * `assoc-nx`
+  * `assoc-nx!`
+* `io.jesi.backpack.closer/close` multimethod
+* `io.jesi.backpack.collection/distinct-vals?`
+* `io.jesi.backpack.miscellaneous/re-quote`
```
Comment (AndreTheHunter) [inline CHANGELOG.md:12]: I did stuff

File: src/io/jesi/backpack/miscellaneous.cljc [hunk 15]
```diff
@@ -44,3 +46,14 @@
         (not (and x y))))
   ([x y & more]
    (xor (xor x y) (apply xor more))))
+
+;see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#escaping
+;regex literal does not work here (wants more \\\), but works in lumo...
```
Comment (AndreTheHunter) [inline src/io/jesi/backpack/miscellaneous.cljc:51]: FML

File: test/io/jesi/backpack/cache_test.cljc [hunk 16]
```diff
@@ -59,10 +59,9 @@
               (reset! captor nil)
               (let [timeout (+ 100 ttl)
                     f #(assert-invoked)]
-                #?(:cljs (js/setTimeout #(do (f) (done)) timeout)
-                   :clj  (do
-                           (Thread/sleep timeout)
-                           (f)))))))))
+                #?@(:cljs [(js/setTimeout #(do (f) (done)) timeout)]
+                    :clj  [(Thread/sleep timeout)
+                           (f)])))))))
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/cache_test.cljc:64]: Got rid of "redundant do" warning

File: test/io/jesi/backpack/miscellaneous_test.cljc [hunk 17]
```diff
@@ -68,14 +69,13 @@
...
-      #?(:clj  (do
-                 (is= expected (bp/->uri expected))
+      #?@(:clj  [(is= expected (bp/->uri expected))
                  (is= expected (bp/->uri uri))
-                 (is (identical? expected (bp/->uri expected))))
-         :cljs (is= uri
-                    (-> uri bp/->uri str)
-                    (-> expected str)
-                    (-> expected bp/->uri str)))
+                 (is (identical? expected (bp/->uri expected)))]
```
Comment (AndreTheHunter) [inline test/io/jesi/backpack/miscellaneous_test.cljc:74]: Got rid of "redundant do" warning

File: src/io/jesi/backpack/collection.cljc (hunk 2 above)
Comment (axrs) [reply src/io/jesi/backpack/collection.cljc:372]: `(comp (partial apply distinct?) vals)`

Comment (axrs) [reply src/io/jesi/backpack/collection.cljc:372]: https://clojuredocs.org/clojure.core/distinct_q

File: src/io/jesi/backpack/macros.cljc [hunk 18]
```diff
@@ -220,3 +222,65 @@ A single default expression can follow the clauses, and its value will be return
...
+    :else
+    (throw (#?(:clj IllegalArgumentException. :cljs js/Error.) "with-open only allows Symbols in bindings"))))
+
+(defmacro with-open-> [x & forms]
+  `(with-open [v# ~x]
+     (-> v#
+         ~@forms)))
+
+(defmacro assoc-nx
+  "Lazily assocs each kv iff its key doesn't already exist.
```
Comment (axrs) [inline src/io/jesi/backpack/macros.cljc:256]:
```suggestion
  "Lazily assocs each kv if its key doesn't already exist.
```

**Review by axrs — CHANGES_REQUESTED** [review]
Smol

File: test/io/jesi/backpack/macros_test.cljc (hunk 6 above)
Comment (axrs) [reply test/io/jesi/backpack/macros_test.cljc:358]: hmmm.

File: src/io/jesi/backpack/macros.cljc (hunk 18 above)
Comment (AndreTheHunter) [reply src/io/jesi/backpack/macros.cljc:256]: That what was is the `assoc-nx`'s docstring. I think "iff" mean "if and only if" https://en.wikipedia.org/wiki/If_and_only_if
I'm happy to change it

File: src/io/jesi/backpack/collection.cljc (hunk 2 above)
Comment (AndreTheHunter) [reply src/io/jesi/backpack/collection.cljc]: Looks like the counting a set is faster. See https://github.com/jesims/backpack/pull/83/files#diff-7eb79c9927f780b45f29cfed61267af6f19f32d7ca022a28dc661cd8b0949430R885

File: src/io/jesi/backpack/macros.cljc (hunk 18 above)
Comment (axrs) [reply src/io/jesi/backpack/macros.cljc]: Nah, that's fine.

**Review by axrs — APPROVED** [review]

**Pull request merged**